librbd: switch IO path to use new librados asio API

dillaman opened the pull request:
Switch the librbd IO path to natively use the librbd::AsioEngine execution context. This enables the use of multiple client threads for IO operations.

<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

mattbenjamin commented:
I love it when a plan comes together

dillaman commented:
> I love it when a plan comes together

still have my doubts about this new librados_asio API -- at quick glance (1) it's still creating these heavy-weight librados::AioCompletion objects, (2) it seems to be completing them via a `finisher_strand` which would be nice to opt-out of, and (3) it starts its own thread-pool instead of allowing you to pass in your own execution context.

adamemerson commented:
@jdillaman The new RADOS API should let you do all of that. No AioCompletion, no threads down to asyncmessenger, and you can supply any execution context you like.

adamemerson commented:
librados_asio is a bridge between ASIO and RGW put in as a stopgap. The new API hat you probably want to use is the one in include/neorados/RADOS.hpp

You also get fewer allocations that way.

dillaman commented:
>  no threads down to asyncmessenger

So the async messenger no longer uses its own `Worker` threads under `neorados`?

> The new API hat you probably want to use is the one in include/neorados/RADOS.hpp

Thanks -- I guess that's what I get by using the current state of RGW as a guide. Since this is in `include/neorados`, does this imply that the API is going to be stable? I don't see it getting pulled into librados-devel, so I presume the answer is no (which I am fine with). 

I see how I can pass my own `io_context` down which is nice. Can I also bootstrap up `RADOS` from a legacy `librados::Rados`-provided `CephContext` via `RADOS::make_with_cct`? The side-effect being that now I will have two connections to the cluster?

adamemerson commented:
@jdillaman No. The messenger still has its own worker threads, unfortunately. I haven't taken a crack at an ASIO messenger, though it would be a nice thing to do.

But you don't need any other thread pools besides whatever you supply for ASIO and the Messenger workers.

Yes, if you have an existing librados::Rados, then you can get its CephContext out and build a neorados::RADOS with it. (And you'll have two connections to the cluster if you keep both around.)

It's not stable yet. At some point I'd like to stabilize it, but I want to get some experience with using it in RGW, RBD, and anywhere else it cam be used before doing so.

dillaman commented:
A few test failures that need investigation but mostly clean: https://pulpito.ceph.com/jdillaman-2020-07-13_18:23:39-rbd-wip-jd-testing-distro-basic-smithi/

trociny commented on src/librbd/image/OpenRequest.cc:
```diff
@@ -482,6 +482,7 @@ Context *OpenRequest<I>::handle_v2_get_data_pool(int *result) {
       }
       m_image_ctx->data_ctx.close();
     } else {
+      m_image_ctx->rebuild_data_io_context();
```
@dillaman `rebuild_data_io_context` uses the current `data_ctx` namespace, but it is set below. Is it intentional?

dillaman replied:
Negative -- will fix in next push

dillaman commented:
https://pulpito.ceph.com/jdillaman-2020-07-14_20:11:06-rbd-wip-jd-testing-distro-basic-smithi/

trociny commented:
@dillaman it needs rebase after #36077 is merged

trociny commented:
Also, please note there is a librbd unittest test failure on jenkins, that might be related:
```
[ RUN      ] TestInternal.ShrinkFlushesCache
/home/jenkins-build/build/workspace/ceph-pull-requests/src/test/librbd/test_internal.cc:1049: Failure
Value of: c->is_complete()
  Actual: false
Expected: true
pure virtual method called
terminate called without an active exception
/home/jenkins-build/build/workspace/ceph-pull-requests/src/test/run-rbd-unit-tests.sh: line 18: 12940 Aborted                 (core dumped) unittest_librbd
```

dillaman commented:
rebased (and dropped offending test case since it really doesn't do a good job testing that the shrink issues a flush as compared to the mock test case)
